pydatastructs/linear_data_structures/linked_lists.py

```diff
@@ -722,3 +724,213 @@ def extract(self, index):
         elif index == 0:
             self.tail.next = self.head
         return node
+
+class SkipList(LinkedList):
+    """
+    Represents Skip List
+
+    Examples
+    ========
+
+    >>> from pydatastructs import SkipList
+    >>> skip_list = SkipList()
+    >>> skip_list.insert(2, "2")
+    >>> skip_list.insert(4, "4")
+    >>> skip_list.insert(6, "4")
+    >>> skip_list.insert(4, "5")
+    >>> skip_list.insert(8, "4")
+    >>> skip_list.insert(9, "4")
+    >>> skip_list.delete(4)
+
+    References
+    ==========
+
+    .. [1] https://epaperpress.com/sortsearch/download/skiplist.pdf
+    .. [2] https://github.com/TheAlgorithms/Python/blob/a9f73e318cddf43769083614a3e1f9dab1ec50fc/data_structures/linked_list/skip_list.py
+
+    Parameters
+    ==========
+
+    p
+        Fraction of nodes of level i which are present at level i+1 also.
+
+    max_level
+        Maximum level that can be attained by any node.
+
+    """
+
+    __slots__ = ['size', 'level', 'p', 'max_level']
+
+    def __new__(cls, p=0.5, max_level=16):
```

Comment:
```suggestion
    def __new__(cls, p, levels):
```

Reply:
max_level is how much skip list can extend vertically it is different from how many actual levels are there

Reply:
In the implementations I have seen, there is always some form of default value for frequency of elements in i+1 level. 0.5 was used in MIT lecture.

Reply:
`p=0.5` is mostly used for perfect skip lists. Let's leave it upto to the programmer to decide what value they want to use.

Reply:
so no default value for p?

Reply:
> max_level is how much skip list can extend vertically it is different from how many actual levels are there

[This](https://epaperpress.com/sortsearch/download/skiplist.pdf) defines MaxLevel as L(N) where N is the upper bound on number of elements which we don't know at the start. Can we do something like providing an option to the user to provide the maximum number of elements that their skip list can have and calculate max levels on the basis of that.

Reply:
we can do that for sure

Reply:
https://github.com/geertj/pyskiplist/blob/c5f94cf135d42bb277255150d3f570ed807468b2/pyskiplist/skiplist.py#L115 in here also they have given max_level by default. I guess the main reason is if max level increases more than log(N) where N is the number of elements, the cost will become worse than binary tree. Increasing the max_level will also mean increasing duplication and since we dont know before hand how many elements are supposed to be there, its better that programmer himself assigns the value.

Reply:
Other thing we can do is that if the max_level is given then take that as max_level else if list of elements is given then max_level= log(N) else take the default value of 16.

Reply:
> > max_level is how much skip list can extend vertically it is different from how many actual levels are there
> 
> [This](https://epaperpress.com/sortsearch/download/skiplist.pdf) defines MaxLevel as L(N) where N is the upper bound on number of elements which we don't know at the start. Can we do something like providing an option to the user to provide the maximum number of elements that their skip list can have and calculate max levels on the basis of that.

Shall I add a parameter for expected number of elements?